src/adr/0013-precompute-jmap-preview.md

```diff
@@ -0,0 +1,45 @@
+# 13. Precompute JMAP Email preview
+
+Date: 2019-10-09
+
+## Status
+
+Accepted (lazy consensus)
+
+## Context
+
+JMAP messages have a handy preview property displaying the firsts 256 characters of meaningful test of a message.
+
+This property is often displayed for message listing in JMAP clients, thus it is queried a lot.
+
+Currently, to get the preview, James retrieves the full message body, parse it using MIME persers, removes HTML and keep meaningful text. This process is expensive, especially for clients relying on polling.
+
+## Decision
+
+We should pre-compute message preview.
+
+A MailboxListener will compute the preview and store it in a MessagePreviewStore.
+
+We should have a Cassandra and memory implementation.
+
+When the preview is precomputed then for these messages we can consider the "preview" property as a metadata.
+
+We should provide a webAdmin task allowing to rebuild the projection.
+
+Some performance tests will be run in order to evaluate the improvements.
+
+## Consequences
+
+We expect a huge performance enhancement for JMAP clients relying on preview for listing mails.
+
+Messages whose preview is not yet computed will still require live preview computation as a fallback mechanism.
```

Comment:
do you intend to store the result of this live preview?

Reply:
IMO yes we should